### DISCUSSION TITLE
Manage own metadata as a user

### DISCUSSION THREAD

**Discussion by dimaip** [zitadel/zitadel#5227] [Q&A]
Hi!
Zitadel newbie here!

I want to be able to store and update some custom user profile fields, and metadata seems like the thing I might want to use.
I'm trying to figure out how to set/update a metadata field as a user.

I tried doing smth like this:

```bash
curl --location --request POST 'http://localhost:8080/management/v1/users/201236897748484098/metadata/asdf' \
--header 'accept: application/json' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer xxx' \
--data-raw '{
  "id": "201236897748484098",
  "key": "asdf",
  "value": "test"
}'
```

With my user's token, I get `"No matching permissions found (AUTH-5mWD2)"`.

If I use the admin's token, I get `"User could not be found (COMMAND-uXHNj)"`, even though the user id is correct.

Could you folks please lend me a hand with this?

**Comment by hifabienne** [comment]
Hei @dimaip 
We do have different APIs for different use cases.
The management API is for admins that have permissions to mange data of ZITADEL on different levels.
If you like to change the data of your own user you have to use the authentication api.
https://zitadel.com/docs/apis/introduction#authentication

We know it is a little bit confusing, thats why we try to rewrite our API docs currently to help you on that side.

Should look something like this:

````
curl --request POST \
  --url http://localhost:8080/auth/v1/users/me/metadata/key \
  --header 'Authorization: Bearer {TOKEN} \
  --header 'Content-Type: application/json' \
  --data '{
	"value": "SGFsbMO2w7ZsZQ=="
}'
```

**Reply by hifabienne** [reply]
Oh and make sure to encode / decode your key with base64

**Reply by dimaip** [reply]
Wow @hifabienne, thanks so much for such prompt reply!

This yields me `Method Not Allowed`.

```bash
curl --location --request POST 'http://localhost:8080/auth/v1/users/me/metadata/asdf' \
--header 'accept: application/json' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer ***' \
--data-raw '{
  "value": "SGFsbMO2w7ZsZQ=="
}'
```


Looking at the API specs, there're only these endpoints available:

<img width="711" alt="image" src="https://user-images.githubusercontent.com/837032/219292151-05fb6405-e5f1-4a10-b916-17940a6768f1.png">


And yes, my original request starts to work if I add `Org User Manager` role to my user, but that's obviously not what I want, since that would allow the user to modify other users.

**Reply by hifabienne** [reply]
Oh yes you are right. At the moment we do not have an endpoint to allow adding your own metadata. I think because we didn't want users to allow just adding random values.

What is your use case?

I currently think of if you are building your own users profile page you want them to be able to add additional predefined data. e.g hobby, departmenent, what ever you need?

**Reply by dimaip** [reply]
Yep!
I have a simple web app, and I'm too lazy to roll my own backend/db for storing some extra fields on users.
Ideally, the user should only be able to post to pre-defined fields, but that's not super important.

One thing I'm considering is that I could possibly create a privileged system account which would have the `Org User Management` role and create a simple lambda function that would take a request from the user and proxy it to the management api, while checking the permissions (basically that it's the right field on its own user).
Of course, it would be great if there was a way to avoid doing that, but I think it should get me going.

**Reply by hifabienne** [reply]
Yes I agree.
We currently just got to the same point where we would need it.
We probably add the possibility in the auth api with this issue, if you like to subscribe: https://github.com/zitadel/zitadel/issues/5213

**Reply by dimaip** [reply]
Nice! Thank you for your support!
I'll keep an eye on that issue and hand-roll my hackish solution to circumvent the issue meanwhile.

**Reply by hifabienne** [reply]
Hei @dimaip 
We just realised that there is a little bit more to consider on this.
It is a security concern as every user that has a token would be able to manipulate all the metadata on the user.

I think we first would have to implement something like this:
https://github.com/zitadel/zitadel/issues/4420

Unfortunately we do not yet have a timeline for it.

**Reply by dimaip** [reply]
@hifabienne yes, exactly!
